src/com/ms/silverking/cloud/dht/daemon/storage/convergence/ConvergenceController2.java

```diff
@@ -697,314 +706,320 @@ private void fetchKeys() {
-        Log.warning("cancelAllOngoingConvergence()");
-        for (ConvergenceController2 cc2 : activeConvergenceControllers) {
-            cc2.cancelConvergence();
-        }
-        activeConvergenceControllers.clear();
-        convergenceControllers.clear();
-        pausedChecksumTrees.clear();
-        pausedSignalledControllers.clear();
+  }
+
+  public static void pauseConvergence() {
+    convergencePauseMutex.lock();
+    try {
+      if (!convergencePaused) {
+        Log.warning("Pausing convergence");
+        convergencePaused = true;
+        convergencePauseCV.signalAll();
+      }
+    } finally {
+      convergencePauseMutex.unlock();
     }
-    
-    ///////////////////////////////////////////////////////////////////////////
-    
-    private static boolean    convergencePaused;
-    private static Lock        convergencePauseMutex = new ReentrantLock();
-    private static Condition    convergencePauseCV = convergencePauseMutex.newCondition();
-    
-    private static BlockingQueue<PausedChecksumTree>        pausedChecksumTrees = new LinkedBlockingQueue<>();
-    private static BlockingQueue<ConvergenceController2>    pausedSignalledControllers = new LinkedBlockingQueue<>();
-    
-    //private static final int    queuePauseThreshold = 100;
-    //private static final int    queueUnpauseThreshold = 50;
-    private static final int    queuePauseThreshold = 50;
-    private static final int    queueUnpauseThreshold = 25;
-    
-    static class PausedChecksumTree {
-        final UUIDBase                     uuid;
-        final ChecksumNode                remoteTree; 
-        final ConvergencePoint            cp; 
-        final MessageGroupConnection    connection;
-        
-        PausedChecksumTree(UUIDBase uuid, ChecksumNode remoteTree, 
-                ConvergencePoint cp, 
-                MessageGroupConnection connection) {
-            this.uuid = uuid;
-            this.remoteTree = remoteTree; 
-            this.cp = cp;
-            this.connection = connection;
-        }
+  }
+
+  public static void unpauseConvergence() {
+    convergencePauseMutex.lock();
+    try {
+      if (convergencePaused) {
+        Log.warning("Unpausing convergence");
+        convergencePaused = false;
+        convergencePauseCV.signalAll();
+      }
+    } finally {
+      convergencePauseMutex.unlock();
     }
-    
-    public static void pauseConvergence() {
-        convergencePauseMutex.lock();
+  }
+
+  private static void blockWhilePaused() {
+    convergencePauseMutex.lock();
+    try {
+      while (convergencePaused) {
         try {
-            if (!convergencePaused) {
-                Log.warning("Pausing convergence");
-                convergencePaused = true;
-                convergencePauseCV.signalAll();
-            }
-        } finally {
-            convergencePauseMutex.unlock();
+          convergencePauseCV.await();
+        } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)